Pdist naming alt

pearsonca opened the pull request:
## Description

@seabbs indicated dissatisfaction with having to pass around pdist/dprimary name arguments. This PR provides an alternative implementation to that approach: using `attr` on the functions and introspecting distribution from function bodies (looking for the base R `Call...` distribution invocations - should be stable naming, as those are public API for R).

This sets approach at soft deprecation, so is approximately as repetitious / noisy code-wise as current approach. When actually deprecated, this solution is much trimmer - all the deprecation checks, arguments, and associated conditional support for them can be removed.

If this PR becomes main, happy to stage up the next incremental revisions dealing with stages of deprecation removal. That ... could make sense? not sure how annoying the fast-forwards would be versus doing it from scratch.

## Checklist

- [X] I have read the [contribution guidelines](https://github.com/epinowcast/.github/blob/main/CONTRIBUTING.md).
- [X] I have tested my changes locally.
- [X] I have added or updated unit tests where necessary.
- [x] I have updated the documentation if required.
- [x] My code follows the established coding standards.
- [x] I have added a news item linked to this PR.
- [x] I have reviewed CI checks for this PR and addressed them as far as I am able.

pearsonca commented on tests/testthat/test-pcens.R:
```diff
@@ -1,21 +1,20 @@
   expect_s3_class(obj, "pcens_pgamma_dunif")
-  expect_identical(obj$pdist, pgamma)
-  expect_identical(obj$dprimary, dunif)
+  expect_identical(body(obj$pdist), body(pgamma))
+  expect_identical(formals(obj$pdist), formals(pgamma))
+  expect_identical(body(obj$dprimary), body(dunif))
+  expect_identical(formals(obj$dprimary), formals(dunif))
```
n.b. expect_identical tests attributes, which no longer match, hence need to change this test.

seabbs replied:
seems to be fine to keep

pearsonca commented on R/pcens.R:
```diff
@@ -13,13 +13,13 @@
-    pdist_name <- .extract_function_name(substitute(pdist))
+    ...) {
+
+  if (is.null(attr(pdist, "name"))) {
+    attr(pdist, "name") <- .extract_function_name(pdist)
   }
```
a possibly desirable alternative: push the conditional check into .extract_function_name - since it no longer depends on NSE name parsing, could make this more like `pdist <- .ensure_function_name(pdist)` and push the check / attribute setting into the utility function.

that would encapsulate the use of attributes, and concentrate future handling of other, non-`stats` distribution sources (somewhat - also need to deal w/ how the attr is *used* elsewhere in the code)

pearsonca replied:
could also make it be side-effect-y / pass-by-reference style, and have `.ensure_function_name(...)` modify items in parent environment, if there are any performance issues here with having to copy-modify-reassign or if that sort of approach is otherwise preferrable stylistically / maintenance wise.

seabbs commented:
We spoke face to face with the conclusion that we think this is a good approach with the following changes: 

- Add depreciations (probably using the `lifecycle` package)
- Make sure this approach works if accessing/assigning the attribute just once in `new_pcens`.
- Add a helper function for users to assign the attribute externally and test this works with custom user versions of distributions. 

We didn`t discuss who would make these changes. I am happy to do so in a PR targeting this one or very happy to review them if @pearsonca has the time to make them directly.

pearsonca commented:
> We spoke face to face with the conclusion that we think this is a good approach with the following changes:
> 
> * Add depreciations (probably using the `lifecycle` package)
> * Make sure this approach works if accessing/assigning the attribute just once in `new_pcens`.
> * Add a helper function for users to assign the attribute externally and test this works with custom user versions of distributions.
> 
> We didn`t discuss who would make these changes. I am happy to do so in a PR targeting this one or very happy to review them if @pearsonca has the time to make them directly.

done with deprecations; don't recall what second bullet means; have implemented the function, but not a test - can you suggest an appropriate example for custom user version?

seabbs commented on DESCRIPTION:
```diff
@@ -1,6 +1,6 @@
 Package: primarycensored
 Title: Primary Event Censored Distributions
-Version: 1.0.0.1000
+Version: 1.0.1.1000
```
```suggestion
Version: 1.0.0.1000
```

we are just incrementing on release so this can stay as is

seabbs commented on DESCRIPTION:
```diff
@@ -22,6 +22,11 @@ Authors@R:
            comment = c(ORCID = "0000-0001-5782-7330")),
+    person(given = "Carl",
+           family = "Pearson",
+           role = c("ctb"),
```
```suggestion
           role = c("aut"),
```

imo this is a comparable amount of work as from other authors

seabbs commented on R/pprimarycensored.R:
```diff
@@ -99,24 +99,17 @@
     q, pdist, pwindow = 1, D = Inf, dprimary = stats::dunif,
-    dprimary_args = list(), pdist_name = NULL, dprimary_name = NULL, ...) {
+    dprimary_args = list(),
+    pdist_name = deprecated(), dprimary_name = deprecated(), ...) {
```
I think this is a hard depreciation but we need a soft one i.e. where the functionality continues to work for the user but with a warning for at least a version

seabbs replied:
> `r lifecycle::badge("deprecated")` this argument will be
#' ignored; use [attach_distribution_name()] on `dprimary` instead

This is what makes me think its a hard dep

pearsonca replied:
i'll switch to slower deprecation protocol.

seabbs commented on R/pcens.R:
```diff
@@ -13,29 +13,18 @@
+    pdist_name = deprecated(), dprimary_name = deprecated(),
+    ...) {
 
+  .name_deprecation(pdist_name, dprimary_name)
```
as in other comments I think this is effectively a hard depreciation but we need to keep the functionality working for at least one release

seabbs commented on tests/testthat/test-pcens.R:
```diff
@@ -1,21 +1,20 @@
 test_that("new_pcens creates object with correct structure", {
-  pdist_name <- "pgamma"
```
in terms of depreciation we basically need to keep this test and show it still works with a depreciation

pearsonca replied:
added a test to the end which is explicitly dealing with the deprecation aspect; i *think* it covers relevant cases, and folks can focus updating that as with the upcoming deprecation stages

seabbs reviewed (changes requested):
This is getting there and the core functionality looks good. As noted I think the depreciation needs to be soft (i.e. old functionality works with a warning) vs doesn't work. Happy to hear push back on this (i.e. they wont get breaks it just wont use the analytical solutions any more and maybe that is okay).

Something it does need however is some tests of the new attribution assignment and recovery. Both on the unit test tevel and integration tests (can I put rnorm in some other function and it all still works - answer yes but test!)

pearsonca commented on R/fitdistdoublecens.R:
```diff
@@ -59,10 +59,11 @@
                               pwindow = 1, D = Inf,
                               dprimary = stats::dunif,
-                              dprimary_name = NULL,
+                              dprimary_name = deprecated(),
```
my read of the lifecycle docs indicates no defaults for when hard deprecation occurs. the docs suggest a deprecation plan (_soft => _warn => _hard => deletion) over a series of versions. i can adjust to work more like this.

could add such to the detail message in the deprecation invocation - up to you.

seabbs replied:
yup that is my understanding as well. The depreciation stuff I forwarded from EpiNow matches what I know of best lifecycle practice

seabbs commented:
All looks good to me. Thanks again for this its a nice change.

pearsonca commented:
> I did some light doc clean up, a review of the use of depreciation, added a test for the attribution functionality both at a unit and integration level and sorted the export/pkgdown issues.
> 
> This looks good to go for me so just waiting on the CI to run.
> 
> Note I changed the name of the attribution function to be really on the nose. I think this needs another pass as for anyone aside from a real power user I think it will be very unclear how or when to use it. Planning on reviewing this when a lookup is added as this may be a helpful wa y to guide users

Re the attribute naming, agree - would be ideal here is focused practical vignette showing its use. you mentioned live use in some other epinowcast repo? perhaps showing an example derived from that.

pearsonca commented:
hmm, re switching to `lifecycle::deprecated()` - i'm not particularly fussed there, but i should highlight that lifecycle recommends the import deprecate (and only deprecate, not the _soft, _warn, _stop functions) vice the `::` style.

my understanding of their take is basically trying to 1) clearly indicate to users (e.g. `deprecated()` is a very clear value) while 2) not leaking implementation details for how deprecation works (e.g. not noising up the docs with what package is doing it).
